src/cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -838,6 +844,170 @@ static PyObject *log_level(PyObject *, PyObject *args) {
     Py_RETURN_NONE;
 }
 
+class GpiClock {
+  public:
+    GpiClock(GpiObjHdl *clk_sig) : clk_signal(clk_sig) {}
+
+    ~GpiClock() { stop(); }
+
+    // Start the clock. Returns nonzero in case of failure:
+    //  - EBUSY if the clock was already started (stop first)
+    //  - EINVAL if the parameters are invalid
+    //  - EAGAIN if registering the toggle callback failed
+    int start(uint64_t period_steps, uint64_t high_steps, bool start_high);
+
+    int stop();
+
+  private:
+    GpiObjHdl *clk_signal = nullptr;
+    GpiCbHdl *clk_toggle_cb_hdl = nullptr;
+
+    uint64_t period = 0;
+    uint64_t t_high = 0;
+
+    int clk_val = 0;
+
+    int toggle();
+    static int toggle_cb(void *gpi_clk);
+};
+
+int GpiClock::start(uint64_t period_steps, uint64_t high_steps,
+                    bool start_high) {
+    if (clk_toggle_cb_hdl) {
+        LOG_ERROR("Failed to start clock: already started -- stop first");
```

Comment:
I don't think these error log messages should be here. The Python exceptions should cover the ones in `start()` at least.

@ktbarrett thoughts?

Reply:
Makes sense. I left it for subsequent toggles in case it fails there, otherwise the clock would just silently die, but removed them for start